rfcs/20210115-tfx-periodic-training.md

```diff
@@ -0,0 +1,445 @@
+    untouched, without launching a new pipeline run.
+
+Overruns are an abnormal condition and should trigger an alert to the pipeline
+owner. If overrun happens frequently, one should consider optimizing the
+pipeline or adjusting the scheduling interval. To recover from an overrun, the
+killed or skipped runs can be manually backfilled using a specified Span id. Or,
+if a skipped run is unrecoverable, a [rolling window](#rolling-window) Resolver
+can be designed to tolerate one or more missing spans.
+
+If the periodic scheduling system lacks an overrun policy, switching to
+asynchronous pipeline operation (with clearly defined behavior via
+[Advanced DSL Semantics RFC](https://github.com/tensorflow/community/blob/master/rfcs/20200601-tfx-udsl-semantics.md))
+is preferable to inventing an ad hoc mechanism to prevent concurrent runs.
+
+Backfill run might also cause concurrency with the routine periodic run or
+another backfill. The pipeline support team should make sure the normal
+scheduling logic is disabled and ensure existing pipeline runs are finished
+before starting backfill. After this, backfilling one run at a time is always
+safe. Parallel backfills can be safe in some cases, e.g., for
+[partial pipelines](#partial-pipeline-execution) when the pipeline owner is
+certain that there is no cross dependency between the updated ML Metadata of the
+parallel runs.
+
+## Future work
+
+As the proposed solution in this doc is based on a batch pipeline, all input
+data required by the pipeline run should be available before the run starts.
+Using small batch pipelines to mimic streaming for faster data ingestion is not
+recommended for the following reasons:
+
+*   Data ingestion (ExampleGen) frequency is tied to training (Trainer)
+    frequency in synchronous components execution mode (the mode currently
+    available in TFX).
+*   Scheduling pipeline runs too frequently might result in
+    [concurrent runs](#concurrent-runs) and cause unexpected behaviour.
+
+Asynchronous component execution is a better fit to keep training at a certain
+frequency (e.g., daily) but have a faster data ingestion (e.g., hourly). And
+async orchestration semantics can also provide better automation and exception
+handling (e.g., for automatic backfill of late-arriving data, or automatic
+recovery if a pipeline component stalls) compared to cron scheduling. We will
+introduce asynchronous component execution in the future.
+
+Some future work items:
+
+*   Integrate with Google Cloud AI Platform, e.g., utilize
+    [Pub/Sub](https://cloud.google.com/pubsub) and
+    [Beam Streaming](https://beam.apache.org/documentation/sdks/python-streaming/)
+    for input data generation, and use
+    [Cloud Functions](https://cloud.google.com/functions) for auto triggering of
+    TFX pipelines.
+*   Support partial pipeline execution.
+*   Trainer warm-start support for Keras.
+*   Asynchronous execution for components.
+*   Custom ExampleGen guidance.
```

Comment:
Special interest on QueryBased

Reply:
WIP! This doc uses file based as examples, but the concept is not tied to file based